### ISSUE TITLE
Async reporter hooks

### ISSUE THREAD

**Issue by ksheedlo** [jasmine/jasmine#842]
See angular/protractor#1938

We want to be able to respond asynchronously to test failures. For instance, taking a screenshot of the browser in selenium webdriver when an end-to-end test fails. We can't do this in `afterEach` or `afterAll` since those functions don't have access to the pass/fail state of the spec. Nor can we do it reliably in the reporter, since `specDone` and `suiteDone` assume synchronous execution.

To address this problem, an optional `done` callback could be added to reporter functions. Here's an example of how we could apply it to `specDone`:

``` js
jasmine.getEnv().addReporter({
  specDone: function (result, done) {
    if (result.failedExpectations.length === 0) {
      return done();
    }
    browser.driver.takeScreenshot(function (data) {
      // write out screenshot data to a file
      done();
    });
  }
});
```

The behavior would be similar to an async spec in that if the reporter hook takes 1 or no arguments, it is assumed to be synchronous; but if there are 2 arguments, it becomes an asynchronous reporter hook.

Would any owners be interested in accepting this as a PR? If so, I'll look into crafting it.

**Comment by slackersoft** [comment]
I'm still on the fence on this issue.

On the one hand, I totally understand what you want to do and how this would enable more powerful reporters that want to do async things.

On the other hand, so far jasmine has been able to treat reporter hooks as fire and forget type callbacks, and this would change that.

It seems that the keeping the fire and forget nature of reporter hooks is really nice, especially for the jasmine internals. And I'm not sure what the break-even point is for this. Thoughts?

**Comment by ksheedlo** [comment]
Totally agree with you that keeping the reporter hooks fire-and-forget would be really nice for jasmine internals. I don't know where the break-even point is either; this is one of the things I was hoping for guidance with. It's going to be hard to tell before any code is written, which is why I'm leaning towards trying it anyway and sending you some code to look at.

One nice thing Jasmine has going for it here is that the hooks are each only called in one place.

```
┌─[ken][Kens-MacBook-Pro][±][master ✓][2.1.2][~/ksheedlo/jasmine]
└─▪ for m in jasmineStarted suiteStarted specStarted specDone suiteDone jasmineDone; do grasp -r -e '__.'$m'(_$)' src/; done
src/core/Env.js:225-227:(multiline):
      reporter.jasmineStarted({
        totalSpecsDefined: totalSpecsDefined
      });
src/core/Env.js:210:          reporter.suiteStarted(suite.result);
src/core/Env.js:370:        reporter.specStarted(spec.result);
src/core/Env.js:364:        reporter.specDone(result);
src/core/Env.js:217:          reporter.suiteDone(result);
```

This will help make it easier to tell where asynchronous behavior needs to propagate. I've only skimmed through so far, but it's at least going to have to propagate through the `TreeProcessor`.

**Comment by slackersoft** [comment]
I think it would be ideal if the async stuff could be kept contained to the `ReportDispatcher`, but I don't know if that is really possible.

**Comment by mvndaai** [comment]
I support making it Async. There are a lot of issues with it being synchronous, that mostly come up in the protractor community https://github.com/angular/protractor/issues/1938.

**Comment by slackersoft** [comment]
Probably the nicest way to get something like this out of the current codebase is to convert the report dispatcher to use `QueueRunner` to handle the report callbacks. This would allow synchronous and asynchronous reporter callback, and should allow the reporter interface to largely stay the same for existing synchronous reporter callbacks.

I would be happy to review a pull request that accomplishes this. If it turns out we would need to change the reporter interface, it may need to wait until we're ready for a 3.0 release.

**Comment by squidfunk** [comment]
+1

**Comment by c4simba** [comment]
+1

**Comment by stowns** [comment]
My personal use-case is that I'd like to write a reporter that optional ships the results to our reporting system via REST api. In the jasmineDone() step i'd like to be able to make several rest calls (which are composed via promises). However, the process exits midway through the calls being made.

**Comment by CommanderXL** [comment]
+1

**Comment by yjaaidi** [comment]
Hello,

As you can see in our fork of `protractor-beautiful-reporter` [https://github.com/wishtack/protractor-beautiful-reporter/blob/master/index.js](https://github.com/wishtack/protractor-beautiful-reporter/blob/master/index.js) we are using the following pattern:

```javascript
class Reporter {

    _asyncFlow: Promise<any>;
    
    jasmineStarted() {

        /* Wait for async tasks triggered by `specDone`. */
        beforeEach(async () => {
            
            await this._asyncFlow;
            this._asyncFlow = null;
            
        });
        
    }
    
    specDone(result) {
        
        this._asyncFlow = this._asyncSpecDone(result);
        
    }
    
    async _asyncSpecDone(result) {
        
        // @todo: Do your async stuff here depending on `result.status`, take screenshots etc...
        // await takeScreenshot();
        
    }

}
```

**Comment by ajaysuwalka** [comment]
@stowns Man, I don't get the words to thank but thanks a lot... You rock.

**Comment by slackersoft** [comment]
This feature has been implemented in the 3.0-features branch and will be released as part of Jasmine 3.0.

Closing this. Thanks for using Jasmine!
